lib/client.js

```diff
@@ -643,90 +643,116 @@ MatrixClient.prototype.setRoomEncryption = function(roomId, config) {
+                if (devices.hasOwnProperty(deviceId)) {
+                    var keys = devices[deviceId];
+                    var key = keys.keys["curve25519:" + deviceId];
+                    if (key == this._olmDevice.deviceCurve25519Key) {
+                        continue;
+                    }
+                    if (!this.sessionStore.getEndToEndSessions(key)) {
+                        devicesWithoutSession.push([userId, deviceId, key]);
                     }
                 }
             }
         }
-        var deferred = q.defer();
-        if (devicesWithoutSession.length > 0) {
-            var queries = {};
-            for (i = 0; i < devicesWithoutSession.length; ++i) {
-                var device = devicesWithoutSession[i];
-                var query = queries[device[0]] || {};
-                queries[device[0]] = query;
-                query[device[1]] = "curve25519";
-            }
-            var path = "/keys/claim";
-            var content = {one_time_keys: queries};
-            var self = this;
-            this._http.authedRequestWithPrefix(
-                undefined, "POST", path, undefined, content,
-                httpApi.PREFIX_UNSTABLE
-            ).done(function(res) {
-                var missing = {};
-                for (i = 0; i < devicesWithoutSession.length; ++i) {
-                    var device = devicesWithoutSession[i];
-                    var userRes = res.one_time_keys[device[0]] || {};
-                    var deviceRes = userRes[device[1]];
-                    var oneTimeKey;
-                    for (var keyId in deviceRes) {
-                        if (keyId.indexOf("curve25519:") === 0) {
-                            oneTimeKey = deviceRes[keyId];
-                        }
-                    }
-                    if (oneTimeKey) {
-                        var sid = self._olmDevice.createOutboundSession(
-                            device[2], oneTimeKey
-                        );
-                        console.log("Started new sessionid " + sid +
-                                    " for device " + device[2]);
-                    } else {
-                        missing[device[0]] = missing[device[0]] || [];
-                        missing[device[0]].push([device[1]]);
-                    }
+    }
+
+    if (devicesWithoutSession.length === 0) {
+        return q({
+            missingUsers: userWithoutDevices,
+            missingDevices: []
+        });
+    }
+
+    var queries = {};
+    for (i = 0; i < devicesWithoutSession.length; ++i) {
+        var device = devicesWithoutSession[i];
+        var query = queries[device[0]] || {};
+        queries[device[0]] = query;
+        query[device[1]] = "curve25519";
+    }
+    var path = "/keys/claim";
+    var content = {one_time_keys: queries};
+    var self = this;
+    return this._http.authedRequestWithPrefix(
+        undefined, "POST", path, undefined, content,
+        httpApi.PREFIX_UNSTABLE
```

Comment:
We can always grep for PREFIX_UNSTABLE, so we think this would be a bit redundant.